File: src/edu/uq/workways/ioporlet/parcoords/ParCoords.java
package edu.uq.workways.ioporlet.parcoords;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.vaadin.annotations.JavaScript;
import com.vaadin.annotations.StyleSheet;
import com.vaadin.ui.AbstractJavaScriptComponent;
@JavaScript({"js/d3.v3.js", "js/d3.parcoords.js","js/ParCoords.js"})
@StyleSheet("js/parcoords.css")
public class ParCoords extends AbstractJavaScriptComponent{
	private static final long serialVersionUID = 1L;

	public ParCoords(String _containerId){
		getState().containerId = _containerId;
		this.setStyleName("parcoords");		
		getState().data = new ArrayList<Map<String, Object>>();
		this.setId(_containerId);
	}
	
	public ParCoords(String _containerId, List<Map<String, Object>> _data){
		getState().containerId = _containerId;
		getState().data = _data;
		this.setStyleName("parcoords");
		this.setId(_containerId);
	}
	
	public void setDimentions(String[] _dimentions){
		getState().dimensions = _dimentions;
	}
	
	public void setDimensionTitles(String[] _dimensionTitles){
		getState().dimensionTitles = _dimensionTitles;
	}
	
	public void setTypes(Map<String, String> _types){
		getState().types = _types;
	}
	
	public void setMargin(Map<String, Integer> _margin){
		getState().margin = _margin;
	}
    
	/**
	 *  public boolean 						brushed 		= false;
	    public String 						mode 			= "default";
	    public int 							rate			= 20;
	    public int 							width			= 600;
	    public int 							height			= 300;
	    public String 						color			= "#069";
	    public String 						composite		= "source-over";
	    public double 						alpha			= 0.7;
	 */
	public void setProperties(String prop, Object value){
	   if (prop.equalsIgnoreCase("brushed")) {
	    getState().brushed=(Boolean)value;
	   } else if (prop.equalsIgnoreCase("mode")) {
	    getState().mode=(String)value;
	   } else if (prop.equalsIgnoreCase("rate")) {
	    getState().rate=(Integer)value;
	   } else if (prop.equalsIgnoreCase("pcwidth")) {
		getState().pcWidth=(Integer)value;
	   } else if (prop.equalsIgnoreCase("pcheight")) {
	    getState().pcHeight=(Integer)value;
	   } else if (prop.equalsIgnoreCase("color")) {
	    getState().color=(String)value;
	   } else if (prop.equalsIgnoreCase("composite")) {
		    getState().composite=(String)value;
	   } else if (prop.equalsIgnoreCase("alpha")) {
		    getState().alpha=(Double)value;
	   }
	   
	}
	
	public void addData(Map<String, Object> _dat){
		getState().data.add(_dat);
	}
	
	@Override
	  protected ParCoordsState getState() {
	    return (ParCoordsState) super.getState();
	  }
	
}

File: src/edu/uq/workways/ioporlet/parcoords/js/ParCoords.js
edu_uq_workways_ioporlet_parcoords_ParCoords = function() {
	var e = this.getElement();
    var parcoords = d3.parcoords()("#"+ this.getState().containerId)
					.alpha(this.getState().alpha)
					.mode(this.getState().mode)
					.width(this.getState().pcWidth)
					.height(this.getState().pcHeight)
					.shadows()
					.autoscale()
					.mode("queue")
					.brushable()
					;
    //first time paint
    var first = true;
	this.onStateChange = function() {
		//data
	  	if(typeof this.getState().data!== 'undefined' && this.getState().data.length > 0){
	  		parcoords.data(this.getState().data);
	  	}
	  	//dimension
	  	if(typeof this.getState().dimensions!== 'undefined' && this.getState().dimensions.length > 0){
	  		parcoords.dimensions(this.getState().dimensions);
	  	}
	  	//types
	  	if(typeof this.getState().types!== 'undefined' && this.getState().types.length > 0){
	  		parcoords.types(this.getState().types);
	  	}
	  	//margin
	  	if(typeof this.getState().margin!== 'undefined' && this.getState().margin.length > 0){
	  		parcoords.margin(this.getState().margin);
	  	}
	  	if(first=== true){
            parcoords.render().createAxes();
            if(this.getState().brushed==true)
            	parcoords.brushable();
            first = false;
	    }
	    else{
            parcoords.render().updateAxes();
	    }
	}
      
};